### PR TITLE
fix: coverage for libraries

### LINKED ISSUE #2567: Coverage: Library coverage
Libraries are currently not covered and they are somewhat non-trivial to map instructions for.

Generally, there are two types of library functions:

- **Internal functions** are inlined in the contract that use them
- **External functions** are called via `delegatecall`

The way coverage works currently is in a few passes:

- First, we detect any sections of code that we want to include in our coverage report
- Second, we also take the base contracts of a contract into account, since all behavior of base contracts are inherited. We need this to map instructions to the items we generated in the first pass using source maps
- Third, we use the source maps of *concrete* contracts to find instructions that mark the items gathered in our first and second step as covered. One item might be covered by multiple instructions in different contracts because of inheritance

The issue with adding library coverage is that the Solidity compiler does not give us any easily digestable information about what libraries are used in contracts. This is in contrast to base contracts which are easily identified using a property of contract AST nodes called `linearizedBaseContracts`.

In order to reliably find what libraries are used in contracts so we can map instructions onto coverage items present in libraries, we need to walk each node in the AST to find call expression nodes.

These call expression nodes include some crude type information, and in the case of library calls, the type information will show up as e.g. `type(library LibraryName)`.

After parsing this information, we need to map the library name back onto an AST node ID so we can find what coverage items are present in the library. The rest of the analysis step in `forge coverage` uses AST node IDs to refer to contracts.

Some unknowns:

- Is the type information correct, i.e. does it take shadowing and import aliases into account?
- Are library names from the type information easily mappable onto AST node IDs corresponding to the library's AST?

### LINKED ISSUE #4654: Should `forge coverage` skip libraries?
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (6c1eee9 2023-03-27T00:12:53.527941Z)

### What command(s) is the bug in?

forge coverage

### Operating System

macOS (Apple Silicon)

### Describe the bug

It's quite alarming to see as the first thing, when code is fully fuzzed and tests are passing:
```
| File                          | % Lines      | % Statements | % Branches   | % Funcs     |
|-------------------------------|--------------|--------------|--------------|-------------|
| library.sol | 0.00% (0/42) | 0.00% (0/54) | 0.00% (0/28) | 0.00% (0/4) |
| Total                         | 0.00% (0/42) | 0.00% (0/54) | 0.00% (0/28) | 0.00% (0/4) |
```

Developer has to turn on `RUST_LOG=warn` and dig back to this: https://github.com/foundry-rs/foundry/pull/2504 , to find out that `forge coverage` doesn't work with libraries.

Maybe the above example should show `(0/0)` throughout the `library.sol` row and have another row above "Total" saying "Skipped" that would have the original `0.00% (0/42) | 0.00% (0/54) | 0.00% (0/28) | 0.00% (0/4)`

### LINKED ISSUE #4854: Coverage doesn't cover internal libs when passing implicit storage vars via the `using` syntax
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (8f246e0 2023-04-28T00:14:27.758082000Z)

### What command(s) is the bug in?

forge coverage --report lcov && genhtml -o forge-coverage lcov.info

### Operating System

macOS (Apple Silicon)

### Describe the bug

When importing an internal lib and utilizing it with the `using` syntax, coverage does not detect any hits on the functions in the library.

Example is shown below (some code omitted for brevity):

```js
import "./lib/StakeManagerLibs.sol";

contract StakeManager is IStakeManager, Initializable {
    ...
    using LibStakesManager for Stakes;

    function stakeFor(uint256 id, uint256 amount) external {
        require(id != 0 && id <= _chains.counter, "INVALID_ID");
        MATIC.safeTransferFrom(msg.sender, address(this), amount);
        _stakes.addStake(msg.sender, id, amount); // <========= Coverage registers no hits here
        ISupernetManager manager = managerOf(id);
        manager.onStake(msg.sender, amount);
        emit StakeAdded(id, msg.sender, amount);
    }


```

Yields the following coverage report:

Screen Shot 2023-04-29 at 7 33 30 PM

However, if I change the syntax of how the library function is called, where the storage variable isn't implicitly passed as the first argument, the coverage is correct.

```js
    function stakeFor(uint256 id, uint256 amount) external {
        require(id != 0 && id <= _chains.counter, "INVALID_ID");
        MATIC.safeTransferFrom(msg.sender, address(this), amount);
        LibStakesManager.addStake(_stakes, msg.sender, id, amount); // <======= Use the library explicitly
        // _stakes.addStake(msg.sender, id, amount);
        ISupernetManager manager = managerOf(id);
        manager.onStake(msg.sender, amount);
        emit StakeAdded(id, msg.sender, amount);
    }
```

Then coverage report is accurate:

Screen Shot 2023-04-29 at 7 36 41 PM

Another important piece of information to add is that calling the Library function explicitly just once, ie the remainder of the `_stakes.addStake(...)` and any additional library calls to any/all functions using the implicit syntax are now covered.

### LINKED ISSUE #7054: forge coverage for libraries is not updating
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (a5efe4f 2024-02-08T00:16:00.244105000Z)

### What command(s) is the bug in?

forge coverage

### Operating System

macOS (Apple Silicon)

### Describe the bug

I am writing unit tests for some libraries and am getting 0% coverage with forge coverage. Feel free to look at this branch as an example! https://github.com/Uniswap/relayer/pull/57

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes #7054
Closes #2567
Closes #4854
Closes #4654

## Solution

Currently we're collecting anchors via the following steps:
1. Walk the AST for each contract definition to find coverage items contained in it along with its base contracts and potentially libraries.
2. Collect mapping from contract_id to coverage items contained in it itself or its base contracts/libraries.
3. Try to find anchors for each coverage item found in step 2.

This approach requires us to be able to reliably find libraries used by a contract which turned out to be non-trivial.

One possible approach could be to reuse some of #6936 code to walk through typed AST and find all contracts/libs used by a given contract. However, this might be breaking due to typed AST not working for some solc versions and this approach is also [not perfect](https://github.com/foundry-rs/foundry/pull/6936#issuecomment-1921443885).

Instead, we now just collect all coverage items and then create a mapping from `source_id` to all coverage items given source contains.

After that, when looking for anchors, we only try finding anchors for items from source ids appearing in source map. This allows us to get rid of dependency resolution through AST.

## Example

I was mainly testing those on Uniswap/v4-core contracts.

Report via latest nightly: [report_nightly.txt](https://github.com/foundry-rs/foundry/files/14783218/report_nightly.txt)
Report via this PR: [report_pr.txt](https://github.com/foundry-rs/foundry/files/14783219/report_pr.txt)

diff:
```diff
9,11c9,11
< | src/libraries/BitMath.sol                  | 0.00% (0/47)      | 0.00% (0/57)       | 0.00% (0/36)     | 0.00% (0/2)      |
< | src/libraries/FullMath.sol                 | 0.00% (0/29)      | 0.00% (0/33)       | 0.00% (0/8)      | 0.00% (0/2)      |
< | src/libraries/Hooks.sol                    | 0.00% (0/41)      | 0.00% (0/76)       | 0.00% (0/28)     | 0.00% (0/14)     |
---
> | src/libraries/BitMath.sol                  | 100.00% (47/47)   | 100.00% (57/57)    | 100.00% (36/36)  | 100.00% (2/2)    |
> | src/libraries/FullMath.sol                 | 100.00% (29/29)   | 100.00% (33/33)    | 100.00% (8/8)    | 100.00% (2/2)    |
> | src/libraries/Hooks.sol                    | 100.00% (41/41)   | 98.68% (75/76)     | 96.43% (27/28)   | 100.00% (14/14)  |
14,23c14,23
< | src/libraries/Pool.sol                     | 0.00% (0/130)     | 0.00% (0/141)      | 0.00% (0/86)     | 0.00% (0/13)     |
< | src/libraries/PoolGetters.sol              | 0.00% (0/2)       | 0.00% (0/2)        | 100.00% (0/0)    | 0.00% (0/2)      |
< | src/libraries/Position.sol                 | 0.00% (0/12)      | 0.00% (0/14)       | 0.00% (0/6)      | 0.00% (0/2)      |
< | src/libraries/SafeCast.sol                 | 0.00% (0/8)       | 0.00% (0/14)       | 0.00% (0/8)      | 0.00% (0/4)      |
< | src/libraries/SqrtPriceMath.sol            | 0.00% (0/32)      | 0.00% (0/66)       | 0.00% (0/24)     | 0.00% (0/8)      |
< | src/libraries/SwapFeeLibrary.sol           | 0.00% (0/5)       | 0.00% (0/9)        | 0.00% (0/4)      | 0.00% (0/3)      |
< | src/libraries/SwapMath.sol                 | 0.00% (0/23)      | 0.00% (0/30)       | 0.00% (0/12)     | 0.00% (0/1)      |
< | src/libraries/TickBitmap.sol               | 0.00% (0/19)      | 0.00% (0/34)       | 0.00% (0/6)      | 0.00% (0/3)      |
< | src/libraries/TickMath.sol                 | 97.87% (92/94)    | 92.57% (137/148)   | 82.61% (38/46)   | 50.00% (2/4)     |
< | src/libraries/UnsafeMath.sol               | 0.00% (0/1)       | 0.00% (0/1)        | 100.00% (0/0)    | 0.00% (0/1)      |
---
> | src/libraries/Pool.sol                     | 96.92% (126/130)  | 94.33% (133/141)   | 89.53% (77/86)   | 100.00% (13/13)  |
> | src/libraries/PoolGetters.sol              | 100.00% (2/2)     | 100.00% (2/2)      | 100.00% (0/0)    | 100.00% (2/2)    |
> | src/libraries/Position.sol                 | 100.00% (12/12)   | 100.00% (14/14)    | 100.00% (6/6)    | 100.00% (2/2)    |
> | src/libraries/SafeCast.sol                 | 100.00% (8/8)     | 100.00% (14/14)    | 100.00% (8/8)    | 100.00% (4/4)    |
> | src/libraries/SqrtPriceMath.sol            | 100.00% (32/32)   | 98.48% (65/66)     | 83.33% (20/24)   | 100.00% (8/8)    |
> | src/libraries/SwapFeeLibrary.sol           | 100.00% (5/5)     | 100.00% (9/9)      | 100.00% (4/4)    | 100.00% (3/3)    |
> | src/libraries/SwapMath.sol                 | 100.00% (23/23)   | 100.00% (30/30)    | 100.00% (12/12)  | 100.00% (1/1)    |
> | src/libraries/TickBitmap.sol               | 100.00% (19/19)   | 100.00% (34/34)    | 100.00% (6/6)    | 100.00% (3/3)    |
> | src/libraries/TickMath.sol                 | 97.87% (92/94)    | 97.30% (144/148)   | 97.83% (45/46)   | 50.00% (2/4)     |
> | src/libraries/UnsafeMath.sol               | 100.00% (1/1)     | 100.00% (1/1)      | 100.00% (0/0)    | 100.00% (1/1)    |
51,53c51,53
< | src/types/BalanceDelta.sol                 | 0.00% (0/2)       | 0.00% (0/2)        | 100.00% (0/0)    | 0.00% (0/2)      |
< | src/types/Currency.sol                     | 0.00% (0/15)      | 0.00% (0/24)       | 0.00% (0/10)     | 0.00% (0/6)      |
< | src/types/PoolId.sol                       | 0.00% (0/1)       | 0.00% (0/2)        | 100.00% (0/0)    | 0.00% (0/1)      |
---
> | src/types/BalanceDelta.sol                 | 100.00% (2/2)     | 100.00% (2/2)      | 100.00% (0/0)    | 100.00% (2/2)    |
> | src/types/Currency.sol                     | 100.00% (15/15)   | 91.67% (22/24)     | 80.00% (8/10)    | 100.00% (6/6)    |
> | src/types/PoolId.sol                       | 100.00% (1/1)     | 100.00% (2/2)      | 100.00% (0/0)    | 100.00% (1/1)    |
55c55
< | test/utils/Deployers.sol                   | 0.00% (0/50)      | 0.00% (0/63)       | 0.00% (0/10)     | 0.00% (0/11)     |
---
> | test/utils/Deployers.sol                   | 96.00% (48/50)    | 93.65% (59/63)     | 80.00% (8/10)    | 90.91% (10/11)   |
58,59c58,59
< | test/utils/SortTokens.sol                  | 0.00% (0/3)       | 0.00% (0/5)        | 0.00% (0/2)      | 0.00% (0/1)      |
< | Total                                      | 43.42% (554/1276) | 42.89% (745/1737)  | 22.24% (153/688) | 40.35% (115/285) |
---
> | test/utils/SortTokens.sol                  | 66.67% (2/3)      | 80.00% (4/5)       | 50.00% (1/2)     | 100.00% (1/1)    |
> | Total                                      | 75.78% (967/1276) | 75.30% (1308/1737) | 55.38% (381/688) | 66.67% (190/285) |
```

cc @onbjerg

### CODE REVIEW

File: crates/evm/coverage/src/analysis.rs [hunk 1]
```diff
@@ -511,127 +466,36 @@ impl SourceAnalyzer {
...
-                            })
-                            .collect(),
-                    );
-                }
-
-                // For tests and contracts with no items we still record an empty Vec so we don't
-                // end up here again
-                if items.is_empty() || is_test {
-                    self.contract_items.insert(contract_id.clone(), Vec::new());
+            let ContractVisitor { items, .. } = ContractVisitor::new(
```
Comment (onbjerg) [inline crates/evm/coverage/src/analysis.rs:470]: w/o the `if !self.contract_items.contains_key(contract_id) {` check, will this not lead to re-analyzing contracts? and possibly duping coverage items?

**Review by onbjerg — COMMENTED** [review]
looks great overall, really stoked, an initial q tho

File: crates/evm/coverage/src/analysis.rs [hunk 2]
```diff
@@ -511,127 +466,36 @@ impl SourceAnalyzer {
...
+                contract_id.source_id,
+                self.sources.get(&contract_id.source_id).unwrap_or_else(|| {
+                    panic!("We should have the source code for source ID {}", contract_id.source_id)
```
Comment (mattsse) [inline crates/evm/coverage/src/analysis.rs:473]: can we return an error?
the function returns an error

File: crates/forge/bin/cmd/coverage.rs [hunk 3]
```diff
@@ -257,19 +257,33 @@ impl CoverageArgs {
...
+                .iter()
+                .enumerate()
+                .map(|(item_id, item)| (item.loc.source_id, item_id))
+                .fold(
+                    HashMap::new(),
+                    |mut map: HashMap<usize, Vec<usize>>, (source_id, item_id)| {
+                        map.entry(source_id).or_default().push(item_id);
+                        map
+                    },
+                );
```
Comment (mattsse) [inline crates/forge/bin/cmd/coverage.rs:273]: here I'd prefer a for loop

**Review by mattsse — CHANGES_REQUESTED** [review]
love to see it.

only a few style nits, defer to @onbjerg 

is there a few to add a few sanity tests for this? probably difficult?

File: crates/evm/coverage/src/analysis.rs (hunk 1 above)
Comment (klkvr) [reply crates/evm/coverage/src/analysis.rs:470]: We're walking only `ContractDefinition` nodes from `self.contracts`, so it should be fine as long as there is no duplicating `ContractId` keys

File: crates/forge/bin/cmd/coverage.rs [hunk 4]
```diff
@@ -257,19 +257,26 @@ impl CoverageArgs {
                 })?,
             )?
             .analyze()?;
-            let anchors: HashMap<ContractId, Vec<ItemAnchor>> = source_analysis
-                .contract_items
+
+            // Build helper mapping used by `find_anchors`
+            let mut items_by_source_id = HashMap::new();
```
Comment (mattsse) [inline crates/forge/bin/cmd/coverage.rs:262]:
```suggestion
            let mut items_by_source_id = HashMap::with_capacity(source_analysis.items.len());
```

File: crates/forge/bin/cmd/coverage.rs [hunk 5]
```diff
@@ -257,19 +257,26 @@ impl CoverageArgs {
...
+
+            for (item_id, item) in source_analysis.items.iter().enumerate() {
+                items_by_source_id.entry(item.loc.source_id).or_insert_with(Vec::new).push(item_id);
```
Comment (mattsse) [inline crates/forge/bin/cmd/coverage.rs:265]: there should be `or_default()` I believe

**Review by mattsse — APPROVED** [review]
lgtm, pending @onbjerg

File: crates/evm/coverage/src/analysis.rs (hunk 1 above)
Comment (onbjerg) [reply crates/evm/coverage/src/analysis.rs:465]: i think thats a fair assumption

**Review by onbjerg — APPROVED** [review]

**Comment by frontier159** [comment]
This sounds great, thanks @klkvr 

It seems to have a regression for different `--report` arguments though:
1. If `--report summary` is explicitly used, it shows zero coverage for everything (should be the same as not specifying according to the --help)
2. If `--report lcov` is used, it shows a lot less coverage than before the change

FWIW, I run via this flow to get a html representation of the coverage and branches:
```
forge coverage --report summary --report lcov
lcov --remove lcov.info 'contracts/test/*' 'test/*' --output-file lcov.info --rc lcov_branch_coverage=1
genhtml lcov.info -o report --branch-coverage --legend
```

**Comment by frontier159** [comment]
Hmm @klkvr  it might have been that I just needed to `forge clean`, after a full refresh it now seems ok. Although github actions also showed the issue on a clean repo. I'll keep an eye on it and let you know if it happens again.

**Comment by frontier159** [comment]
Hmm it might be non-deterministic, I just tried again and it failed this time. Odd

You can try [this repo](https://github.com/TempleDAO/origami-public/):
1. clone with submodules
2. nvm use
3. yarn
4. yarn forge-coverage-html

**Pull request merged**